source/s2n/s2n_tls_channel_handler.c

```diff
@@ -717,6 +719,38 @@ static int s_parse_protocol_preferences(
     return AWS_OP_SUCCESS;
 }
 
+static size_t s_tl_cleanup_key = 0; /* Address of variable serves as key in hash table */
+static struct aws_event_loop_local_object s_tl_cleanup_object;
+
+static void s_aws_cleanup_s2n_thread_local_state(void *user_data) {
+    (void)user_data;
+
+    aws_tls_clean_up_thread_local_state();
+}
+
+static int s_s2n_tls_channel_handler_schedule_tl_cleanup(struct aws_channel_slot *slot) {
+    struct aws_channel *channel = slot->channel;
+
+    struct aws_event_loop_local_object existing_marker;
+    AWS_ZERO_STRUCT(existing_marker);
+
+    if (aws_channel_fetch_local_object(channel, &s_tl_cleanup_key, &existing_marker)) {
+        /* Doesn't exist in event loop table: add it and add the at-exit cleanup callback */
+        AWS_ZERO_STRUCT(s_tl_cleanup_object);
+        s_tl_cleanup_object.key = &s_tl_cleanup_key;
+        s_tl_cleanup_object.object = NULL;
+        s_tl_cleanup_object.on_object_removed = NULL;
```

Comment:
Since we're inserting the same object into multiple threads, there's a chance that on thread zeroes out memory while another thread is reading it.

You could just set all this at static initialization time
```
static struct aws_event_loop_local_object s_tl_cleanup_object = {
    .key = &s_tl_cleanup_key,
    ...
};
```